pyrival/algebra/discrete_log.py

```diff
@@ -1,22 +1,50 @@
+from math import gcd, sqrt
+
+
+def extended_gcd(a, b):
+    if not b:
+        return a, 1, 0
+
+    g, y, x = extended_gcd(b, a % b)
+    return g, x, y - ((a // b) * x)
+
+
 def discrete_log(a, b, m):
-    """returns some x s.t. pow(a, x, m) == b or None if no such x exists, works when a amd m are coprime"""
-    a %= m
-    b %= m
-    if b == 1:
+    """returns smallest non-negative x s.t. pow(a, x, m) == b or None if no such x exists"""
+    if b >= m:
+        return None
+
+    if b == 1 or m == 1:
         return 0
 
-    n, vals = int(m**0.5) + 1, {}
-    mult, cur = pow(a, n, m), 1
+    i, t = 1, a
+    while (1 << i) <= m:
```

Comment:
Store `2**i` rather than `i` in a variable and multiply by 2 in each loop.

Reply:
What advantages can we gain storing both? If speed is your concern then firstly this loop will be executed `O(log m)` times at most, where the main complexity of discrete log lies below, i.e `O(sqrt(m))`. And secondly it's not necessarily true storing in another variable will be faster than a simple bit shift with `int`, and even if it were it'd be negligible.

Reply:
Sure, that is reasonable.